src/buffer/out/textBuffer.cpp

```diff
@@ -166,6 +166,29 @@ ROW& TextBuffer::_getRowByOffsetDirect(size_t offset)
     return *reinterpret_cast<ROW*>(row);
 }
 
+// This function is "direct" because it trusts the caller to know that this row
+// could be the scratchpad row.
+size_t TextBuffer::_estimateOffsetOfLastCommittedRowDirect() const noexcept
+{
+    if (_commitWatermark == _buffer.get())
+    {
+        // Estimation failed. Act as though we have at least one non-scratchpad row committed.
+        // The caller will be passing this offset back to GetRowByOffset or similar later, which
+        // will commit it.
+        return 1;
+    }
+    const auto lastRowOffset = ((_commitWatermark - 1) - _buffer.get()) / _bufferRowStride;
+    return gsl::narrow_cast<size_t>(lastRowOffset);
+}
+
+// Returns the "user-visible" index of the last committed row, which can be used
+// to short-circuit some algorithms that try to scan the entire buffer..
+size_t TextBuffer::_estimateOffsetOfLastCommittedRow() const noexcept

```

Comment:
I think it would make sense to return `til::CoordType` because that's what we use elsewhere for coordinates and offsets and so no further casting is needed.

Reply:
meh, yeah, you're right